### PR TITLE
sql package lint + minor efficiencies

### PR DESCRIPTION
**Pull request by amoore877**
Signed-off-by: Andrew Moore <andrew.s.moore@uber.com>

**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [X] Documentation updated?

**Affected functionality**
No functional changes to API

**Description of change**
While walking through sql plugin to gain understanding, fixed lint errors and made minor efficiency enhancements. Example: registration entry validation should occur prior to getting the transaction lock.

**Which issue this PR fixes**
Primarily lint problems

### CODE REVIEW

**Comment by amoore877** [comment]
addressed comment on effective comments

File: pkg/server/plugin/datastore/sql/sql.go [hunk 1]
```diff
@@ -93,6 +95,7 @@ func (ds *sqlPlugin) UpdateBundle(ctx context.Context, req *datastore.UpdateBund
 	return resp, nil
 }
 
+// AppendBundle append bundle's certificates to the existing bundle (by trust domain). If no existing one is present, create it.
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]:
```
// AppendBundle appends bundle contents to...
```

File: pkg/server/plugin/datastore/sql/sql.go [hunk 2]
```diff
@@ -148,6 +155,7 @@ func (ds *sqlPlugin) CreateAttestedNode(ctx context.Context,
 	return resp, nil
 }
 
+// FetchAttestedNode fetch an existing attested node by SPIFFE ID
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *fetches*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 3]
```diff
@@ -160,6 +168,7 @@ func (ds *sqlPlugin) FetchAttestedNode(ctx context.Context,
 	return resp, nil
 }
 
+// ListAttestedNodes list all attested nodes (pagination available)
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *lists*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 4]
```diff
@@ -172,6 +181,7 @@ func (ds *sqlPlugin) ListAttestedNodes(ctx context.Context,
 	return resp, nil
 }
 
+// UpdateAttestedNode update the given node's cert serial and expiration.
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *updates*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 5]
```diff
@@ -196,7 +207,12 @@ func (ds *sqlPlugin) DeleteAttestedNode(ctx context.Context,
 	return resp, nil
 }
 
+// SetNodeSelectors set node (agent) selectors by SPIFFE ID, deleting old selectors first
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *sets*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 6]
```diff
@@ -206,6 +222,7 @@ func (ds *sqlPlugin) SetNodeSelectors(ctx context.Context, req *datastore.SetNod
 	return resp, nil
 }
 
+// GetNodeSelectors get node (agent) selectors by SPIFFE ID
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *gets*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 7]
```diff
@@ -230,6 +252,7 @@ func (ds *sqlPlugin) CreateRegistrationEntry(ctx context.Context,
 	return resp, nil
 }
 
+// FetchRegistrationEntry fetch an existing registration by entry ID
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *fetches*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 8]
```diff
@@ -242,6 +265,7 @@ func (ds *sqlPlugin) FetchRegistrationEntry(ctx context.Context,
 	return resp, nil
 }
 
+// ListRegistrationEntries list all registrations (pagination available)
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *lists*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 9]
```diff
@@ -254,8 +278,12 @@ func (ds *sqlPlugin) ListRegistrationEntries(ctx context.Context,
 	return resp, nil
 }
 
-func (ds sqlPlugin) UpdateRegistrationEntry(ctx context.Context,
+// UpdateRegistrationEntry update the existing registration entry
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *updates an*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 10]
```diff
@@ -266,6 +294,7 @@ func (ds sqlPlugin) UpdateRegistrationEntry(ctx context.Context,
 	return resp, nil
 }
 
+// DeleteRegistrationEntry delete the given registration
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *deletes*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 11]
```diff
@@ -301,6 +334,7 @@ func (ds *sqlPlugin) FetchJoinToken(ctx context.Context, req *datastore.FetchJoi
 	return resp, nil
 }
 
+// DeleteJoinToken Delete the given join token
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *deletes*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 12]
```diff
@@ -323,8 +357,8 @@ func (ds *sqlPlugin) PruneJoinTokens(ctx context.Context, req *datastore.PruneJo
 	return resp, nil
 }
 
+// Configure parse HCL config payload into config struct, and open new DB based on the result
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: *parses*
*opens*

File: pkg/server/plugin/datastore/sql/sql.go [hunk 13]
```diff
@@ -524,10 +559,10 @@ func deleteBundle(tx *gorm.DB, req *datastore.DeleteBundleRequest) (*datastore.D
 
 	// Get a count of associated registration entries
 	entriesAssociation := tx.Model(model).Association("FederatedEntries")
-	entriesCount := entriesAssociation.Count()
 	if err := entriesAssociation.Error; err != nil {
 		return nil, sqlError.Wrap(err)
 	}
+	entriesCount := entriesAssociation.Count()
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: The order of statements is important here. Count() is the method that actually executes the query. As such, the error check must come after.

File: pkg/server/plugin/datastore/sql/sql.go [hunk 14]
```diff
@@ -610,17 +645,13 @@ func listBundles(tx *gorm.DB, req *datastore.ListBundlesRequest) (*datastore.Lis
 	return resp, nil
 }
 
+// assumes non-nil node in request
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: With this refactoring, it becomes clear that the the flow for each exported method is to 1) validate input, 2) take TX lock, 3) forward to unexported function to do the work. I don't think this comment is adding much value and risks becoming stale if the exported method does additional validation that would no longer be reflected in this comment.

I'd suggest removing it.

File: pkg/server/plugin/datastore/sql/sql.go [hunk 15]
```diff
@@ -717,10 +748,8 @@ func deleteAttestedNode(tx *gorm.DB, req *datastore.DeleteAttestedNodeRequest) (
 	}, nil
 }
 
+// setNodeSelectors DB call for setting node selectors. assumes request has non-nil selectors
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: I suggest removing (little value, risk of becoming stale).

File: pkg/server/plugin/datastore/sql/sql.go [hunk 16]
```diff
@@ -760,18 +789,11 @@ func getNodeSelectors(tx *gorm.DB, req *datastore.GetNodeSelectorsRequest) (*dat
 	}, nil
 }
 
+// createRegistrationEntry DB call for creating a registration entry. Assumes request has non-nil,
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: I suggest removing (little value, risk of becoming stale).

File: pkg/server/plugin/datastore/sql/sql.go [hunk 17]
```diff
@@ -1000,17 +1022,10 @@ func findRegisteredEntries(entryTx *gorm.DB, p *datastore.Pagination) ([]Registe
 	return entries, p, nil
 }
 
+// updateRegistrationEntry DB call for updating a registration entry. Assumes request's entry is non-nil
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: I suggest removing (little value, risk of becoming stale).

File: pkg/server/plugin/datastore/sql/sql.go [hunk 18]
```diff
@@ -1083,11 +1098,9 @@ func deleteRegistrationEntry(tx *gorm.DB,
 	}, nil
 }
 
+// createJoinToken DB call to create a join token. Assumes request's join token is non-nil and non-empty,
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: I suggest removing (little value, risk of becoming stale).

File: pkg/server/plugin/datastore/sql/sql.go [hunk 19]
```diff
@@ -1150,7 +1163,13 @@ func modelToBundle(model *Bundle) (*datastore.Bundle, error) {
 	return bundle, nil
 }
 
+// validateRegistrationEntry validate the registration entry is non-nil, has >0 selectors,
```
Comment (azdagron) [inline pkg/server/plugin/datastore/sql/sql.go]: Considering this is a non-doc comment, it seems to provide too much detail and risks becoming stale if the function internals change. Validation details seem easy to discern from the function body, which is small and simple.

I'd suggest removing the comment.

File: pkg/server/plugin/datastore/sql/postgres.go [hunk 20]
```diff
@@ -2,6 +2,7 @@ package sql
 
 import (
 	"github.com/jinzhu/gorm"
+	// gorm postgres dialect init registration
```
[the remark this answers is not in the document]
Comment (amoore877) [reply pkg/server/plugin/datastore/sql/postgres.go]: yes; lint is that blank imports must have a comment to justify usage

File: pkg/server/plugin/datastore/sql/sql.go (hunk 14 above)
Comment (amoore877) [reply pkg/server/plugin/datastore/sql/sql.go]: will remove, but my overall concern was if other methods come in later to use this one

**Pull request merged**